File: modules/herdr-ssh/ios/HerdrSshModule.swift
import ExpoModulesCore

/// Connection config as it arrives from JavaScript. Records flatten the TS
/// discriminated union — Expo's Record decoding has no sum type — so `authKind`
/// selects which of the optional fields is meaningful.
struct SshConfigRecord: Record {
  @Field var host: String = ""
  @Field var port: Int = 22
  @Field var username: String = ""
  @Field var authKind: String = "privateKey"
  @Field var password: String?
  @Field var privateKey: String?
  @Field var passphrase: String?
  @Field var hostKeyFingerprint: String?
}

/// SSH transport for reaching herdr hosts, as a TurboModule.
///
/// There is no maintained SSH client for React Native, so this wraps the one
/// the SwiftUI app already proved in production (Citadel / SwiftNIO SSH) and
/// exposes exactly the two operations the app needs: run a command, and follow
/// a long-lived one line by line.
///
/// Nothing here throws for an expected failure. Talking to someone else's
/// machine fails routinely — host down, key rotated, herdr not installed — and
/// each of those means something different to the user. Native returns a tagged
/// result and TypeScript owns the policy, where it can be tested.
public class HerdrSshModule: Module {
  private let connections = ConnectionStore()

  public func definition() -> ModuleDefinition {
    Name("HerdrSsh")

    Events("onStreamLine", "onStreamEnd", "onStreamError")

    AsyncFunction("connect") { (id: String, config: SshConfigRecord) -> [String: Any] in
      do {
        let connection = await self.connections.connection(for: id, config: config)
        try await connection.connected()
        let fingerprint = await connection.acceptedFingerprint ?? ""
        return ["ok": true, "fingerprint": fingerprint]
      } catch let failure as SshFailure {
        await self.connections.drop(id)
        return ["ok": false, "code": failure.code, "message": failure.message]
      } catch {
        await self.connections.drop(id)
        return ["ok": false, "code": "connect_failed", "message": SshFailure.friendly(error)]
      }
    }

    AsyncFunction("disconnect") { (id: String) in
      await self.connections.drop(id)
    }

    AsyncFunction("exec") { (id: String, command: String) -> [String: Any] in
      guard let connection = await self.connections.existing(id) else {
        return Self.notConnected
      }
      do {
        let output = try await connection.exec(command)
        return [
          "ok": true,
          "stdout": output.stdout,
          "stderr": output.stderr,
          "exitCode": output.exitCode,
        ]
      } catch let failure as SshFailure {
        return ["ok": false, "code": failure.code, "message": failure.message]
      } catch {
        return ["ok": false, "code": "transport_failed", "message": SshFailure.friendly(error)]
      }
    }

    AsyncFunction("startStream") { (id: String, streamId: String, command: String) -> [String: Any] in
      guard let connection = await self.connections.existing(id) else {
        return Self.notConnected
      }
      do {
        let task = try await connection.startStream(
          command,
          onLine: { [weak self] line in
            self?.sendEvent("onStreamLine", ["streamId": streamId, "line": line])
          },
          onEnd: { [weak self] exitCode in
            self?.sendEvent("onStreamEnd", ["streamId": streamId, "exitCode": exitCode])
            guard let store = self?.connections else { return }
            Task { await store.forgetStream(streamId) }
          },
          onError: { [weak self] code, message in
            self?.sendEvent("onStreamError", [
              "streamId": streamId, "code": code, "message": message,
            ])
            guard let store = self?.connections else { return }
            Task { await store.forgetStream(streamId) }
          }
        )
        await self.connections.registerStream(streamId, task: task)
        return ["ok": true]
      } catch let failure as SshFailure {
        return ["ok": false, "code": failure.code, "message": failure.message]
      } catch {
        return ["ok": false, "code": "transport_failed", "message": SshFailure.friendly(error)]
      }
    }

    AsyncFunction("stopStream") { (streamId: String) in
      await self.connections.stopStream(streamId)
    }

    OnDestroy {
      let store = self.connections
      Task { await store.closeAll() }
    }
  }

  private static let notConnected: [String: Any] = [
    "ok": false,
    "code": "not_connected",
    "message": "No live connection for this server. Connect first.",
  ]
}

/// Owns the live connections and stream tasks. An actor because JS can call
/// `connect` for two servers, or `stopStream` while a stream is still starting,
/// concurrently.
actor ConnectionStore {
  private var connections: [String: SshConnection] = [:]
  private var streams: [String: Task<Void, Never>] = [:]

  func connection(for id: String, config: SshConfigRecord) -> SshConnection {
    if let existing = connections[id] { return existing }
    let created = SshConnection(config: config)
    connections[id] = created
    return created
  }

  func existing(_ id: String) -> SshConnection? {
    connections[id]
  }

  func registerStream(_ streamId: String, task: Task<Void, Never>) {
    streams[streamId] = task
  }

  func forgetStream(_ streamId: String) {
    streams[streamId] = nil
  }

  func stopStream(_ streamId: String) {
    streams.removeValue(forKey: streamId)?.cancel()
  }

  func drop(_ id: String) async {
    guard let connection = connections.removeValue(forKey: id) else { return }
    await connection.close()
  }

  func closeAll() async {
    for task in streams.values { task.cancel() }
    streams.removeAll()
    for connection in connections.values { await connection.close() }
    connections.removeAll()
  }
}

File: modules/herdr-ssh/android/src/main/java/expo/modules/herdrssh/HerdrSshModule.kt
package expo.modules.herdrssh

import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record
import org.bouncycastle.jce.provider.BouncyCastleProvider
import java.security.Security
import java.util.concurrent.ConcurrentHashMap

/**
 * Android ships a cut-down BouncyCastle under the provider name "BC", and sshj
 * asks for "BC" by name: with Android's copy there is no X25519, no EC and no
 * SHA-256 digest, so every handshake failed with "no such algorithm: X25519
 * for provider BC" and no host could be reached (#4). The Compose app replaced
 * the provider at startup; the Expo port lost that step. Swapping in the full
 * library once, before the first connection, restores it.
 */
internal object Crypto {
  @Volatile private var installed = false

  fun ensureBouncyCastle() {
    if (installed) return
    synchronized(this) {
      if (installed) return
      Security.removeProvider(BouncyCastleProvider.PROVIDER_NAME)
      Security.addProvider(BouncyCastleProvider())
      installed = true
    }
  }
}

/**
 * Connection config as it arrives from JavaScript. Records flatten the TS
 * discriminated union — Expo's Record decoding has no sum type — so `authKind`
 * selects which of the optional fields is meaningful.
 */
class SshConfigRecord : Record {
  @Field var host: String = ""
  @Field var port: Int = 22
  @Field var username: String = ""
  @Field var authKind: String = "privateKey"
  @Field var password: String? = null
  @Field var privateKey: String? = null
  @Field var passphrase: String? = null
  @Field var hostKeyFingerprint: String? = null
}

/**
 * SSH transport for reaching herdr hosts. See `modules/herdr-ssh/README.md` for
 * why this module exists and why nothing here throws for an expected failure.
 */
class HerdrSshModule : Module() {
  private val connections = ConcurrentHashMap<String, SshConnection>()
  private val streams = ConcurrentHashMap<String, SshConnection.StreamHandle>()

  override fun definition() = ModuleDefinition {
    Name("HerdrSsh")

    Events("onStreamLine", "onStreamEnd", "onStreamError")

    OnCreate { Crypto.ensureBouncyCastle() }

    AsyncFunction("connect") Coroutine { id: String, config: SshConfigRecord ->
      try {
        Crypto.ensureBouncyCastle()
        val connection = connections.getOrPut(id) { SshConnection(config) }
        connection.connected()
        mapOf("ok" to true, "fingerprint" to (connection.acceptedFingerprint ?: ""))
      } catch (failure: SshFailure) {
        connections.remove(id)?.close()
        failureMap(failure.code, failure.message)
      } catch (error: Exception) {
        connections.remove(id)?.close()
        failureMap("connect_failed", error.message ?: "Couldn't reach the host.")
      }
    }

    AsyncFunction("disconnect") Coroutine { id: String ->
      connections.remove(id)?.close()
    }

    AsyncFunction("exec") Coroutine { id: String, command: String, timeoutMs: Int ->
      val connection = connections[id] ?: return@Coroutine notConnected
      try {
        val output = connection.exec(command, timeoutMs)
        mapOf(
          "ok" to true,
          "stdout" to output.stdout,
          "stderr" to output.stderr,
          "exitCode" to output.exitCode,
        )
      } catch (failure: SshFailure) {
        failureMap(failure.code, failure.message)
      } catch (error: Exception) {
        failureMap("transport_failed", error.message ?: "The connection dropped.")
      }
    }

    AsyncFunction("startStream") Coroutine {
      id: String, streamId: String, command: String, startTimeoutMs: Int ->
      val connection = connections[id] ?: return@Coroutine notConnected
      try {
        val handle = connection.startStream(
          command,
          startTimeoutMs,
          onLine = { line ->
            sendEvent("onStreamLine", mapOf("streamId" to streamId, "line" to line))
          },
          onEnd = { exitCode ->
            sendEvent("onStreamEnd", mapOf("streamId" to streamId, "exitCode" to exitCode))
            streams.remove(streamId)
          },
          onError = { code, message ->
            sendEvent(
              "onStreamError",
              mapOf("streamId" to streamId, "code" to code, "message" to message),
            )
            streams.remove(streamId)
          },
        )
        streams[streamId] = handle
        mapOf("ok" to true)
      } catch (failure: SshFailure) {
        failureMap(failure.code, failure.message)
      } catch (error: Exception) {
        failureMap("transport_failed", error.message ?: "The stream couldn't start.")
      }
    }

    AsyncFunction("stopStream") Coroutine { streamId: String ->
      streams.remove(streamId)?.stop()
    }

    OnDestroy {
      streams.values.forEach { it.stop() }
      streams.clear()
      // Closed, not just forgotten: clearing the map left every socket open
      // until the process died (#110).
      connections.values.forEach { it.abandon() }
      connections.clear()
    }
  }

  private fun failureMap(code: String, message: String) =
    mapOf("ok" to false, "code" to code, "message" to message)

  private val notConnected = mapOf(
    "ok" to false,
    "code" to "not_connected",
    "message" to "No live connection for this server. Connect first.",
  )
}
